pkg/controller/instance/instance_controller.go

```diff
@@ -141,12 +149,40 @@ func (r *Reconciler) Reconcile(request ctrl.Request) (ctrl.Result, error) {
 		return reconcile.Result{}, err // OV not found has to be retried because it can really have been created after Instance
 	}
 
-	// ---------- 2. First check if we should start execution of new plan ----------
+	// ---------- 2. Check if the object is being deleted, start cleanup plan ----------
 
-	planToBeExecuted, err := instance.GetPlanToBeExecuted(ov)
-	if err != nil {
-		return reconcile.Result{}, err
+	var planToBeExecuted *string
+
+	// A delete request is indicated by a non-zero 'metadata.deletionTimestamp',
+	// see https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#finalizers
+	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
+		if _, hasCleanupPlan := ov.Spec.Plans[kudov1beta1.CleanupPlanName]; hasCleanupPlan {
+			if err := r.addFinalizer(instance); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	} else {
+		log.Printf("InstanceController: Instance %s/%s is being deleted", instance.Namespace, instance.Name)
+		if contains(instance.ObjectMeta.Finalizers, instanceCleanupFinalizerName) {
```

Comment:
I would love if this logic was inside `GetPlanToBeExecuted` that way the controller flow will remain much cleaner and we won't have two branches looking for "plan to be executed"

Reply:
Agreed, the instance is a better fit for these functions.